File: clients/java/cldf-java/src/main/java/io/cldf/examples/CLDFExample.java
package io.cldf.examples;

import java.io.File;
import java.io.IOException;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.Arrays;
import java.util.List;

import io.cldf.api.CLDF;
import io.cldf.api.CLDFArchive;
import io.cldf.models.*;

/** Example usage of the CLDF Java client library. */
public class CLDFExample {

  public static void main(String[] args) {
    try {
      // Create a new CLDF archive
      CLDFArchive archive = createSampleArchive();

      // Write to file
      File outputFile = new File("example.cldf");
      CLDF.write(archive, outputFile);
      System.out.println("CLDF archive written to: " + outputFile.getAbsolutePath());

      // Read from file
      CLDFArchive readArchive = CLDF.read(outputFile);
      System.out.println("\nRead CLDF archive:");
      System.out.println("Version: " + readArchive.getManifest().getVersion());
      System.out.println("Locations: " + readArchive.getLocations().size());
      System.out.println("Climbs: " + readArchive.getClimbs().size());
      System.out.println("Sessions: " + readArchive.getSessions().size());

      // Display some climb data
      System.out.println("\nClimbs:");
      for (Climb climb : readArchive.getClimbs()) {
        System.out.println(
            "- "
                + climb.getRouteName()
                + " ("
                + climb.getGrades().getGrade()
                + ") - "
                + climb.getFinishType());
      }

    } catch (IOException e) {
      System.err.println("Error: " + e.getMessage());
      e.printStackTrace();
    }
  }

  private static CLDFArchive createSampleArchive() {
    // Create manifest
    Manifest manifest =
        Manifest.builder()
            .version("1.0.0")
            .format("CLDF")
            .creationDate(OffsetDateTime.now())
            .appVersion("1.0.0")
            .platform(Manifest.Platform.Desktop)
            .stats(
                Manifest.Stats.builder().locationsCount(2).climbsCount(5).sessionsCount(2).build())
            .build();

    // Create locations
    List<Location> locations =
        Arrays.asList(
            Location.builder()
                .id(1)
                .name("The Spot Bouldering Gym")
                .isIndoor(true)
                .country("USA")
                .state("Colorado")
                .coordinates(
                    Location.Coordinates.builder().latitude(40.0170).longitude(-105.2830).build())
                .terrainType(Location.TerrainType.artificial)
                .starred(true)
                .build(),
            Location.builder()
                .id(2)
                .name("Clear Creek Canyon")
                .isIndoor(false)
                .country("USA")
                .state("Colorado")
                .rockType(Location.RockType.granite)
                .terrainType(Location.TerrainType.natural)
                .accessInfo("Park at pullout mile marker 269")
                .build());

    // Create sessions
    List<Session> sessions =
        Arrays.asList(
            Session.builder()
                .id("session-1")
                .date(LocalDate.now().minusDays(7))
                .location("The Spot Bouldering Gym")
                .locationId("1")
                .isIndoor(true)
                .climbType(Climb.ClimbType.boulder)
                .sessionType(Session.SessionType.indoorBouldering)
                .partners(Arrays.asList("Alice", "Bob"))
                .notes("Great session, worked on crimpy problems")
                .build(),
            Session.builder()
                .id("session-2")
                .date(LocalDate.now().minusDays(1))
                .location("Clear Creek Canyon")
                .locationId("2")
                .isIndoor(false)
                .climbType(Climb.ClimbType.boulder)
                .sessionType(Session.SessionType.bouldering)
                .weather(
                    Session.Weather.builder()
                        .conditions("Sunny")
                        .temperature(22.0)
                        .humidity(30.0)
                        .build())
                .approachTime(15)
                .build());

    // Create climbs
    List<Climb> climbs =
        Arrays.asList(
            Climb.builder()
                .id(1)
                .sessionId(1)
                .date(LocalDate.now().minusDays(7))
                .routeName("Purple Crimps")
                .type(Climb.ClimbType.boulder)
                .finishType(Climb.FinishType.flash)
                .grades(
                    Climb.GradeInfo.builder()
                        .system(Climb.GradeInfo.GradeSystem.vScale)
                        .grade("V4")
                        .build())
                .attempts(1)
                .rating(4)
                .color("#800080")
                .isIndoor(true)
                .tags(Arrays.asList("crimpy", "technical"))
                .build(),
            Climb.builder()
                .id(2)
                .sessionId(1)
                .date(LocalDate.now().minusDays(7))
                .routeName("Yellow Overhang")
                .type(Climb.ClimbType.boulder)
                .finishType(Climb.FinishType.top)
                .grades(
                    Climb.GradeInfo.builder()
                        .system(Climb.GradeInfo.GradeSystem.vScale)
                        .grade("V5")
                        .build())
                .attempts(3)
                .rating(5)
                .color("#FFFF00")
                .isIndoor(true)
                .tags(Arrays.asList("overhang", "powerful"))
                .notes("Finally got the heel hook beta!")
                .build(),
            Climb.builder()
                .id(3)
                .sessionId(2)
                .date(LocalDate.now().minusDays(1))
                .routeName("Warmup Traverse")
                .type(Climb.ClimbType.boulder)
                .finishType(Climb.FinishType.top)
                .grades(
                    Climb.GradeInfo.builder()
                        .system(Climb.GradeInfo.GradeSystem.vScale)
                        .grade("V2")
                        .build())
                .attempts(1)
                .isIndoor(false)
                .rockType(Location.RockType.granite)
                .build(),
            Climb.builder()
                .id(4)
                .sessionId(2)
                .date(LocalDate.now().minusDays(1))
                .routeName("The Egg")
                .type(Climb.ClimbType.boulder)
                .finishType(Climb.FinishType.project)
                .grades(
                    Climb.GradeInfo.builder()
                        .system(Climb.GradeInfo.GradeSystem.vScale)
                        .grade("V7")
                        .build())
                .attempts(5)
                .falls(4)
                .rating(5)
                .isIndoor(false)
                .rockType(Location.RockType.granite)
                .beta("Start matched on sloper, big move to crimp rail")
                .notes("So close! Got to the top hold but couldn't match")
                .build(),
            Climb.builder()
                .id(5)
                .sessionId(2)
                .date(LocalDate.now().minusDays(1))
                .routeName("Classic Arete")
                .type(Climb.ClimbType.boulder)
                .finishType(Climb.FinishType.repeat)
                .grades(
                    Climb.GradeInfo.builder()
                        .system(Climb.GradeInfo.GradeSystem.vScale)
                        .grade("V3")
                        .build())
                .attempts(1)
                .isRepeat(true)
                .rating(4)
                .isIndoor(false)
                .rockType(Location.RockType.granite)
                .tags(Arrays.asList("arete", "technical"))
                .build());

    // Create tags
    List<Tag> tags =
        Arrays.asList(
            Tag.builder()
                .id("tag-1")
                .name("crimpy")
                .isPredefined(true)
                .predefinedTagKey(Tag.PredefinedTagKey.crimpy)
                .color("#FF6B6B")
                .category("holds")
                .build(),
            Tag.builder()
                .id("tag-2")
                .name("overhang")
                .isPredefined(true)
                .predefinedTagKey(Tag.PredefinedTagKey.overhang)
                .color("#4ECDC4")
                .category("angle")
                .build(),
            Tag.builder()
                .id("tag-3")
                .name("project")
                .isPredefined(false)
                .color("#FFD93D")
                .category("custom")
                .build());

    return CLDFArchive.builder()
        .manifest(manifest)
        .locations(locations)
        .sessions(sessions)
        .climbs(climbs)
        .tags(tags)
        .build();
  }
}

File: clients/dart/cldf/example/cldf_example.dart
import 'package:cldf/cldf.dart';

void main() async {
  // Create a sample CLDF archive
  final archive = CLDFArchive(
    manifest: Manifest(
      version: '1.0.0',
      creationDate: DateTime.now(),
      platform: Platform.iOS,
      appVersion: '1.0.0',
      description: 'Sample climbing data export',
      author: Author(name: 'John Climber', email: 'john@example.com'),
    ),
    locations: [
      Location(
        id: 1,
        name: 'Sample Crag',
        country: 'USA',
        isIndoor: false,
        state: 'Colorado',
        city: 'Boulder',
        coordinates: Coordinates(latitude: 40.0150, longitude: -105.2705),
        rockType: RockType.granite,
        terrainType: TerrainType.natural,
      ),
    ],
    routes: [
      Route(
        id: 1,
        locationId: 1,
        name: 'Classic Line',
        routeType: RouteType.route,
        grades: {'french': '6a', 'yds': '5.10a'},
        height: 20,
        bolts: 8,
        qualityRating: 4,
        tags: ['classic', 'well-protected'],
      ),
      Route(
        id: 2,
        locationId: 1,
        name: 'Power Problem',
        routeType: RouteType.boulder,
        grades: {'vScale': 'V4', 'font': '6A'},
        color: '#FF0000',
        tags: ['crimpy', 'overhang'],
      ),
    ],
    climbs: [
      Climb(
        id: 1,
        date: '2024-01-15',
        routeId: 1,
        type: ClimbType.route,
        finishType: FinishType.redpoint,
        attempts: 3,
        belayType: BelayType.lead,
        rating: 4,
        notes: 'Great climb, tricky crux in the middle',
        grades: GradeInfo(
          system: GradeSystem.french,
          grade: '6a',
          conversions: {'yds': '5.10a'},
        ),
      ),
    ],
    sessions: [
      Session(
        id: 1,
        date: '2024-01-15',
        locationId: 1,
        sessionType: SessionType.sportClimbing,
        partners: ['Jane Doe', 'Bob Smith'],
        weather: Weather(conditions: 'Sunny', temperature: 18, humidity: 45),
        notes: 'Perfect conditions for climbing',
      ),
    ],
    tags: [
      Tag(
        id: 1,
        category: 'style',
        isPredefined: false,
        name: 'crimpy',
        description: 'Routes with small crimpy holds',
      ),
      Tag(
        id: 2,
        category: 'style',
        isPredefined: false,
        name: 'overhang',
        description: 'Overhanging routes',
      ),
    ],
  );

  // Write to file
  final writer = CLDFWriter();
  await writer.writeFile('sample.cldf', archive);
  print('Archive written to sample.cldf');

  // Read back the file
  final reader = CLDFReader();
  final loadedArchive = await reader.readFile('sample.cldf');

  print('Loaded archive:');
  print('- Locations: ${loadedArchive.locations.length}');
  print('- Routes: ${loadedArchive.routes?.length ?? 0}');
  print('- Climbs: ${loadedArchive.climbs?.length ?? 0}');
  print('- Sessions: ${loadedArchive.sessions?.length ?? 0}');
  print('- Tags: ${loadedArchive.tags?.length ?? 0}');
}
